From c07e340319ff887ed5b0a39d98c621659e86cdb7 Mon Sep 17 00:00:00 2001
From: Bryce Poole <Bryce.Poole@intel.com>
Date: Thu, 25 Feb 2016 16:03:35 -0800
Subject: [PATCH 355/441] sd card: Added Card Stable support

When the SD Card is inserted/ejected, the CARD_PRESENT bit
is a debounced version of SD_CD# pin (SD Card Detect pin) from
the SD Card.

Adding a wait for the debounce to complete by waiting for just
over one second for CARD_STABLE to be asserted.
---
 drivers/mmc/host/sdhci.c | 27 ++++++++++++++++++++++++++-
 drivers/mmc/host/sdhci.h |  1 +
 2 files changed, 27 insertions(+), 1 deletion(-)

--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -1714,9 +1714,34 @@ static void sdhci_set_ios(struct mmc_hos
 	sdhci_runtime_pm_put(host);
 }
 
+int sdhci_is_card_stable(struct sdhci_host *host)
+{
+	return !!(sdhci_readl(host, SDHCI_PRESENT_STATE) & SDHCI_CARD_IS_STABLE);
+}
+ 
+void sdhci_wait_for_card_stable(struct sdhci_host *host)
+{
+    int cnt = 0;
+
+    while(!sdhci_is_card_stable(host) && cnt < 12)
+    {
+        mdelay(100);
+        cnt++;
+    }
+
+    if(!sdhci_is_card_stable(host))
+    {
+        pr_err("%s: Card Present signal not stable\n", mmc_hostname(host->mmc));
+    }
+}
+
 static int sdhci_do_get_cd(struct sdhci_host *host)
 {
-	int gpio_cd = mmc_gpio_get_cd(host->mmc);
+	int gpio_cd;
+
+	sdhci_wait_for_card_stable(host);
+
+	gpio_cd = mmc_gpio_get_cd(host->mmc);
 
 	if (host->flags & SDHCI_DEVICE_DEAD)
 		return 0;
--- a/drivers/mmc/host/sdhci.h
+++ b/drivers/mmc/host/sdhci.h
@@ -69,6 +69,7 @@
 #define  SDHCI_SPACE_AVAILABLE	0x00000400
 #define  SDHCI_DATA_AVAILABLE	0x00000800
 #define  SDHCI_CARD_PRESENT	0x00010000
+#define  SDHCI_CARD_IS_STABLE	0x00020000
 #define  SDHCI_WRITE_PROTECT	0x00080000
 #define  SDHCI_DATA_LVL_MASK	0x00F00000
 #define   SDHCI_DATA_LVL_SHIFT	20
